cmd/derper/stun.go

```diff
@@ -0,0 +1,170 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"expvar"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"net/netip"
+	"os"
+	"os/exec"
+	"time"
+
+	"tailscale.com/metrics"
+	"tailscale.com/net/stun"
+)
+
+var (
+	stats           = new(metrics.Set)
+	stunDisposition = &metrics.LabelMap{Label: "disposition"}
+	stunAddrFamily  = &metrics.LabelMap{Label: "family"}
+	stunReadError   = stunDisposition.Get("read_error")
+	stunNotSTUN     = stunDisposition.Get("not_stun")
+	stunWriteError  = stunDisposition.Get("write_error")
+	stunSuccess     = stunDisposition.Get("success")
+
+	stunIPv4 = stunAddrFamily.Get("ipv4")
+	stunIPv6 = stunAddrFamily.Get("ipv6")
+)
+
+// statsEntry is the structure of the JSON output of the above stats.
+type statsEntry struct {
+	CounterAddrfamily struct {
+		Ipv4 int64 `json:"ipv4"`
+		Ipv6 int64 `json:"ipv6"`
+	} `json:"counter_addrfamily"`
+	CounterRequests struct {
+		NotStun    int64 `json:"not_stun"`
+		ReadError  int64 `json:"read_error"`
+		Success    int64 `json:"success"`
+		WriteError int64 `json:"write_error"`
+	} `json:"counter_requests"`
+}
+
+func (e *statsEntry) Set() {
+	stunIPv4.Set(e.CounterAddrfamily.Ipv4)
+	stunIPv6.Set(e.CounterAddrfamily.Ipv6)
+	stunNotSTUN.Set(e.CounterRequests.NotStun)
+	stunReadError.Set(e.CounterRequests.ReadError)
+	stunSuccess.Set(e.CounterRequests.Success)
+	stunWriteError.Set(e.CounterRequests.WriteError)
+}
+
+func init() {
+	stats.Set("counter_requests", stunDisposition)
+	stats.Set("counter_addrfamily", stunAddrFamily)
+	expvar.Publish("stun", stats)
+}
+
+// printSTUNStats prints STUN stats to w every d until ctx is done.
+func printSTUNStats(ctx context.Context, w io.Writer, d time.Duration) {
+	ticker := time.NewTicker(d)
+	for {
+		expvar.Do(func(kv expvar.KeyValue) {
+			if kv.Key == "stun" {
+				fmt.Fprintf(w, "%s\n", kv.Value)
+			}
+		})
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+		}
+	}
+}
+
+// readSTUNStats reads lines from r containing STUN statistics and updates matching expvar values.
+func readSTUNStats(ctx context.Context, r io.Reader) {
+	d := json.NewDecoder(r)
+	var entry statsEntry
+	for {
+		if err := d.Decode(&entry); err != nil {
+			return
+		}
+		entry.Set()
+		if ctx.Err() != nil {
+			return
+		}
+	}
+}
+
+// serveChildSTUN starts a stun server in a child process. If the process exits before context is done, serveChildSTUN will with a log entry.
+func startChildSTUN(ctx context.Context) {
+	cmd := exec.Command(os.Args[0], append(os.Args[1:], "-stun-only=true")...)
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		log.Fatalf("stun: failed to create stdout pipe: %v", err)
+	}
+	cmd.Stderr = os.Stderr
+	err = cmd.Start()
+	if err != nil {
+		log.Fatalf("stun: failed to start subprocess: %v", err)
+	}
+	readSTUNStats(ctx, stdout)
+	cmd.Process.Kill()
```

Comment:
Worth noting that this sends a `SIGKILL`; maybe worth allowing things to gracefully shut down with a `SIGTERM` and then waiting and killing with `SIGKILL`?

```golang
	cmd.Process.Signal(syscall.SIGTERM)
	killed := make(chan struct{})
	go func() {
		select {
		case <-killed:
		case <-time.After(2*time.Second):
			cmd.Process.Kill()
		}
	}()
	cmd.Process.Wait()
	close(killed)
```

Reply:
This kill is probably superfluous anyway - if the read has stopped it's more than likely the pipe is closed, which most commonly means the process is gone. I just didn't want to have wait hang indefinitely.